### PR TITLE
[controller-manager] Switch `SecretBinding` controller to `controller-runtime`

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area dev-productivity scalability
/kind enhancement

**What this PR does / why we need it**:
Refactor the `SecretBinding` controller to `controller-runtime`.

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/4251

**Special notes for your reviewer**:
Generally, we want to follow this cookbook while refactoring existing controllers:

1. Add documentation
2. Add integration test based on `envtest`
3. Switch controller to `controller-runtime`

**Release note**:

```breaking dependency
NONE
```

### CODE REVIEW

File: docs/concepts/controller-manager.md [hunk 1]
```diff
@@ -146,6 +146,15 @@ In order to activate it, provide the following configuration:
 > :warning: In addition, you should also configure the `--event-ttl` for the kube-apiserver to define an upper-limit of how long Shoot-related events should be stored.
 The `--event-ttl` should be larger than the `ttlNonShootEvents` or this controller will have no effect.
 
+### [`SecretBinding` Controller](../../pkg/controllermanager/controller/secretbinding)
+
+`SecretBinding`s reference `Secret`s and `Quota`s and are themselves are referenced by `Shoot`s.
+The controller adds finalizers to the referenced objects to ensure they don't get deleted while sitll being referenced.
```
Comment (seshachalam-yv) [inline docs/concepts/controller-manager.md]: 🤔 
```suggestion
The controller adds finalizers to the referenced objects to ensure they don't get deleted while still being referenced.
```

File: docs/concepts/controller-manager.md [hunk 2]
```diff
@@ -146,6 +146,15 @@ In order to activate it, provide the following configuration:
 > :warning: In addition, you should also configure the `--event-ttl` for the kube-apiserver to define an upper-limit of how long Shoot-related events should be stored.
 The `--event-ttl` should be larger than the `ttlNonShootEvents` or this controller will have no effect.
 
+### [`SecretBinding` Controller](../../pkg/controllermanager/controller/secretbinding)
+
+`SecretBinding`s reference `Secret`s and `Quota`s and are themselves are referenced by `Shoot`s.
```
Comment (shafeeqes) [inline docs/concepts/controller-manager.md]:
```suggestion
`SecretBinding`s reference `Secret`s and `Quota`s and are themselves referenced by `Shoot`s.
```

File: pkg/controllermanager/controller/secretbinding/secretbinding_control.go [hunk 3]
```diff
@@ -154,8 +154,8 @@ func (r *secretBindingReconciler) Reconcile(ctx context.Context, request reconci
 		}
 	}
 
-	// Add the Gardener finalizer to the referenced SecretBinding secret to protect it from deletion as long as
-	// the SecretBinding resource does exist.
+	// Add the Gardener finalizer to the referenced Secret secret to protect it from deletion as long as the
```
Comment (shafeeqes) [inline pkg/controllermanager/controller/secretbinding/secretbinding_control.go]:
```suggestion
	// Add the Gardener finalizer to the referenced Secret to protect it from deletion as long as the

```

File: pkg/controllermanager/controller/secretbinding/reconciler.go [hunk 4]
```diff
@@ -18,64 +18,36 @@ import (
 	"context"
```
Comment (shafeeqes) [inline pkg/controllermanager/controller/secretbinding/reconciler.go:18]: Change the year to 2022, as this is a new file now?

**Review by shafeeqes — COMMENTED** [review]
Very nice PR!

**Review by timebertt — COMMENTED** [review]
/approve

**Pull request merged**